### ISSUE TITLE
During video playback on Youtube, press play/pause button rapidly, and the screen blacks out, becoming unresponsive

### ISSUE THREAD

**Issue by npark-mozilla** [mozilla-mobile/firefox-tv#969]
### Steps to reproduce
- Go to youtube, play a video
- While video is playing, press the play/pause button on remote rapidly

### Expected behavior
- It pauses/resumes playback

### Actual behavior
- Shows black screen, play/rewind/FF buttons are unresponsive
- Workaround: show overlay with the option button, then the video resumes

### Device information

* Fire TV device: Fire tv cube
* Firefox version: 2.2 release branch

**Comment by mcomella** [comment]
When reproducing this on youtube and hitting the overlay button to fix the problem, we sometimes see a loading circle or jump to the next video: I wonder if this a Youtube problem where it's trying to do some playback error mitigations and jumps to the next video. 🤔

**Comment by mcomella** [comment]
I can't reproduce on w3c's html5 video page: I think this has something to do with Youtube's player. @npark-mozilla Have you been able to repro anywhere else?

**Comment by mcomella** [comment]
I can't reproduce this with desktop on YouTube TV. I injected:
```js
function pp() { var v = document.querySelector('video'); if (v.paused) v.play(); else v.pause(); }
window.setInterval(pp, 250)
```

I noticed that switching tabs will black out the screen, which might be like opening the overlay? If not, I didn't reproduce it accurately. Also, there seems to be a noticeable delay when pausing the Fire TV (maybe that's because of the remote transmission though?) so maybe I just can't burden my CPU/GPU enough like I can on Fire TV.

**Comment by mcomella** [comment]
Looking at logcat, it looks like the HTMLMediaElement is actually stuck in a play-pause loop. I removed the code that listens for JS playback state and this still happens so I don't think it's our code:
- MediaSession isn't calling us to inject new code
- We don't have JS listeners to execute existing code

Next steps: see if we can reproduce this in a standalone app and on the emulator.

**Comment by mcomella** [comment]
I can reproduce on both v3.0 and v2.2.

I can't reproduce without MediaSession using this code:
```kotlin
    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        val fragmentManager = supportFragmentManager
        val maybeBrowserFragment = (fragmentManager.findFragmentByTag(BrowserFragment.FRAGMENT_TAG) as BrowserFragment?)?.let {
            if (it.isVisible) it else null
        }

        if (event.action != KeyEvent.ACTION_DOWN) {
            return false
        } else {
            maybeBrowserFragment?.webView?.evalJS("var v = document.querySelector('video'); if (v.paused) v.play(); else v.pause();")
            return true
        }
```

**Comment by mcomella** [comment]
I am able to reproduce in MediaSession using similar code to the above and removing the other code we inject:
```kotlin
        override fun onPlay() {
            webView?.evalJS("document.querySelector('video').play();")
            val pb = cachedPlaybackStateBuilder
                    .setState(STATE_PLAYING, PLAYBACK_POSITION_UNKNOWN, 1f)
                    .build()
            mediaSession.setPlaybackState(pb)
        }

        override fun onPause() {
            webView?.evalJS("document.querySelector('video').pause();")
            val pb = cachedPlaybackStateBuilder
                    .setState(STATE_PAUSED, PLAYBACK_POSITION_UNKNOWN, 1f)
                    .build()
            mediaSession.setPlaybackState(pb)
        }
```

Guesses:
- MediaSession is doing something funky to playback state
- We're hitting a memory issue that throws off the native player
- We're leaking (or not leaking) keystrokes to the web content, which is interpreting them differently and creating issues.

**Comment by mcomella** [comment]
I can't reproduce if I send the keystrokes directly to Youtube and let them handle play/pause.

**Comment by mcomella** [comment]
One difference between talking directly with the DOM and MediaSession is that the DOM is switching over it's current state while MediaSession may be sending redundant play/pause events because the playback state hasn't changed quickly enough.

It also seems like I have to hit a very particular cadence to reproduce which is maybe why I can't get it to reproduce using automated methods.
It does seem like I can only reproduce it with a particular

**Comment by mcomella** [comment]
I'm adding log statements. A standard pause is:
```
07-12 14:52:55.380 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:55.452 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:55.691 10042-10042/org.mozilla.tv.firefox.debug D/lol: onPause
07-12 14:52:55.696 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
```

`playPause pressed` is in `dispatchKeyEvent`, `onPause` is the MediaSession requesting a pause, and there's a log statement when the engine gets the pause event.

Here's a full run (first MARK is when the video is displayed, second MARK is when the screen goes black):
```
07-12 14:52:18.446 10042-10042/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 14:52:18.538 10042-10042/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 14:52:19.321 10042-10042/org.mozilla.tv.firefox.debug W/cr_BindingManager: Cannot call determinedVisibility() - never saw a connection for the pid: 10042
07-12 14:52:20.607 10042-10244/org.mozilla.tv.firefox.debug D/ACodec: dataspace changed to 0x10100 (R:0(Unspecified), P:1(BT709_5), M:0(Unspecified), T:0(Unspecified)) (R:0(Unspecified), S:1(BT709), T:0(Unspecified))
07-12 14:52:22.028 10042-10243/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9.bitrateInKbps = 3478
07-12 14:52:22.120 10042-10243/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9 Resolution change from 0x0 to 1920x1080 took = 1
07-12 14:52:23.701 10042-10042/org.mozilla.tv.firefox.debug W/cr_BindingManager: Cannot call determinedVisibility() - never saw a connection for the pid: 10042
07-12 14:52:23.969 10042-10244/org.mozilla.tv.firefox.debug D/ACodec: dataspace changed to 0x10100 (R:0(Unspecified), P:1(BT709_5), M:0(Unspecified), T:0(Unspecified)) (R:0(Unspecified), S:1(BT709), T:0(Unspecified))
07-12 14:52:24.927 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:25.061 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:25.214 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:25.265 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
07-12 14:52:25.313 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
07-12 14:52:25.359 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:25.365 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:amazon_renderer_surface_view_manager.cc(58)] Sending external surface dispose request to browser
07-12 14:52:25.369 10042-10243/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9 Dropping frame
07-12 14:52:25.371 10042-10243/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9 Dropping frame
07-12 14:52:25.371 10042-10237/org.mozilla.tv.firefox.debug W/chromium: [WARNING:audio_sync_reader.cc(180)] ASR: No room in socket buffer.
07-12 14:52:25.372 10042-10243/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9 Dropping frame
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9 Dropping frame
07-12 14:52:25.379 10042-10241/org.mozilla.tv.firefox.debug W/cr_MediaCodecBridge: Releasing: OMX.amlogic.vp9.decoder.awesome
07-12 14:52:25.497 10042-10241/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9.TotalVideoPlaybackTimeMs = 7908
07-12 14:52:25.498 10042-10241/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9.videoResolution.FHD = 7908
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.StreamAvgFramerate = 12.013
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.TwoConsecutiveFramesDropped = 2
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.TotalFramesDropped = 4
07-12 14:52:25.499 10042-10241/org.mozilla.tv.firefox.debug I/MediaCodecLogger: org.mozilla.tv.firefox.debug.4K.HW.video.vp9.MaxConsecutiveFramesDropped = 0
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.ASAP-P5Ms = 130
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.ASAP-P2Ms = 35
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.ASAP-P3Ms = 200
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.FirstFrameReadyMs = 129
    org.mozilla.tv.firefox.debug.4K.HW.video.vp9.StreamAvgBitrateInKbps = 3478
07-12 14:52:25.501 10042-10241/org.mozilla.tv.firefox.debug W/cr_MediaCodecBridge: Codec released
07-12 14:52:25.561 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:25.566 10042-10042/org.mozilla.tv.firefox.debug W/cr_BindingManager: Cannot call determinedVisibility() - never saw a connection for the pid: 10042
07-12 14:52:25.695 10042-10054/org.mozilla.tv.firefox.debug I/art: Background sticky concurrent mark sweep GC freed 84839(4MB) AllocSpace objects, 11(224KB) LOS objects, 22% free, 18MB/23MB, paused 4.271ms total 122.464ms
07-12 14:52:25.721 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:25.854 10042-10042/org.mozilla.tv.firefox.debug D/lol: onPause
07-12 14:52:25.877 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:amazon_renderer_surface_view_manager.cc(63)] Active external surface changed: -1
07-12 14:52:25.953 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
07-12 14:52:26.063 10042-10042/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 14:52:26.229 10042-10042/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
```

What's interesting here is that the video 1) pauses twice and 2) tries to pause and dispose of the surface even before we get on `MediaSessionCallbacks.onPause`: 
```
07-12 14:52:24.927 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:25.061 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:25.214 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 14:52:25.265 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
07-12 14:52:25.313 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:HTMLMediaElement.cpp(2479)] Pause video
07-12 14:52:25.359 10042-10042/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 14:52:25.365 10042-10142/org.mozilla.tv.firefox.debug I/chromium: [INFO:amazon_renderer_surface_view_manager.cc(58)] Sending external surface dispose request to browser
```

Maybe `MediaSession` passes key presses to the system if it receives them too quickly so the WebView is trying to handle them?

**Comment by mcomella** [comment]
In this run, we didn't even receive the MediaSessionCallbacks.onPause before the screen went black:
```
07-12 15:15:26.808 15648-15648/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 15:15:28.374 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 15:15:28.477 15648-15648/org.mozilla.tv.firefox.debug D/lol: onMediaButtonEvent
07-12 15:15:28.489 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 15:15:28.664 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 15:15:28.731 15648-15648/org.mozilla.tv.firefox.debug D/lol: onMediaButtonEvent
07-12 15:15:28.753 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 15:15:28.884 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 15:15:29.011 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 15:15:29.134 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? true
07-12 15:15:29.315 15648-15648/org.mozilla.tv.firefox.debug D/lol: playPause pressed. Is down? false
07-12 15:15:29.596 15648-15648/org.mozilla.tv.firefox.debug D/lol: onMediaButtonEvent
07-12 15:15:29.613 15648-15648/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 15:15:29.740 15648-15648/org.mozilla.tv.firefox.debug D/lol: ----- MARK -----
07-12 15:15:29.897 15648-15648/org.mozilla.tv.firefox.debug D/lol: onPause
```

But we are receiving multiple `onMediaButtonEvents` (presumably this is a synchronous command). Without understanding further, we might be able to workaround the bug by dropping redundant play/pause mediaButtonEvents in `onMediaButtonEvent` (either based on time or on whether on not `onPause` or `onPlay` was received).

**Comment by mcomella** [comment]
Got it: a double tap of play/pause will send a media_next event (I read about this somewhere...). Combined with #963, this creates an infinite loop of media_next events which freezes the browser. With the fix for #963 applied, double-tapping play-pause will just advance to the next video.

Fixed by #963: closing.

fwiw, it was very helpful to actually print out the keys received to fix this issue, rather than just printing out that key strokes were received.
